### PR TITLE
Remove or refactor validation extension function errors

### LINKED ISSUE #193: Improvements to ExtensionsError variants representing internal invariant violations
### Category

User level API changes

### Describe the feature you'd like to request

Several variants of `ExtensionsError` -- `HasNoType`, `FuncMultiplyDefined`, and `MultipleConstructorsSameSignature` -- represent internal invariant violations more than they represent something the user did wrong.  In the current world where we control all the language extensions, these should never be thrown, and if they are thrown, it represents a bug in our code.

We should mark these as internal errors -- perhaps by making a struct `InternalError` (or some other name) and having that be a variant of `ExtensionsError`.  If possible, we should also reorganize `ExtensionsError` and/or `EvaluationError` so that these errors are (a) caught as early as possible, eg before `is_authorized()`, and (b) not part of `EvaluationError`, so that consumers of `EvaluationError` don't have to worry about matching on these.

We want to catch these as early as possible, but we also don't want to do unnecessary expensive checks on the hot path.  One option is to do it when constructing an `Extensions` object, if that is not on the hot path or can be moved off the hot path.

### Describe the solution you'd like

As above.

### Describe alternatives you've considered

Removing these variants entirely in favor of some other mechanism or refactor that makes those errors impossible.  This might be very difficult without restructuring our extensions mechanism.

### Additional context

_No response_

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request
- [ ] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by john-h-kastner-aws**
## Description of changes

Fix #193.

* Remove `WrongCallStyle` error. The corresponding eval error was refactored out a while ago, but the validation error stuck around.
* Remove `MultiplyDefinedFunctionError`. Report this error only once when constructing extension schema.

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [ ] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [ ] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [ ] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

### CODE REVIEW

File: cedar-policy-core/src/extensions.rs [hunk 1]
```diff
@@ -350,6 +350,34 @@ pub mod extension_function_lookup_errors {
 /// Type alias for convenience
 pub type Result<T> = std::result::Result<T, ExtensionFunctionLookupError>;
 
+/// Utilities shared with the `cedar-policy-validator` extensions module.
```
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/extensions.rs:353]: or rather, will be shared when this and #1090  are merged

**Review by shaobo-he-aws — APPROVED** [review]
LGTM.

File: cedar-policy-validator/src/extensions.rs [hunk 2]
```diff
@@ -56,3 +114,29 @@ fn eval_extension_constructor(
         RestrictedExpr::call_extension_fn(constructor_name, [RestrictedExpr::val(lit_str_arg)]);
     evaluator.interpret(constructor_call_expr.as_borrowed())
 }
+
+/// Errors occurring while initializing extensions. These are internal errors, so
+/// this enum should not become part of the public API unless we publicly expose
+/// user-defined extension function.
```
Comment (khieta) [inline cedar-policy-validator/src/extensions.rs:120]:
```suggestion
/// user-defined extension functions.
```

**Review by khieta — APPROVED** [review]
Lgtm.

File: cedar-policy-validator/src/extensions.rs [hunk 3]
```diff
@@ -33,15 +39,67 @@ pub mod decimal;
...
-/// Get schemas for all the available extensions.
-pub fn all_available_extension_schemas() -> Vec<ExtensionSchema> {
-    vec![
+lazy_static::lazy_static! {
+    static ref ALL_AVAILABLE_EXTENSION_SCHEMA_OBJECTS : Vec<ExtensionSchema> = vec![
         #[cfg(feature = "ipaddr")]
         ipaddr::extension_schema(),
         #[cfg(feature = "decimal")]
         decimal::extension_schema(),
         partial_evaluation::extension_schema(),
```
[the remark this answers is not in the document]
Comment (john-h-kastner-aws) [reply cedar-policy-validator/src/extensions.rs:48]: that a good question, and I'm not sure.

It's also not behind the feature for `cedar-policy-core` which means a policy this parses

```cedar
permit(principal, action, resource) when { unknown("foo") };
```

```bash
cargo -q run authorize --entities <(echo '[]') --principal 'P::""' --action 'A::""' --resource 'R::""' <<<'permit(principal, action, resource) when { unknown("foo") };'
```
```
DENY

error while evaluating policy `policy0`: the expression contains unknown(s): `true && (unknown("foo"))`
```

[the remark this answers is not in the document]
Comment (khieta) [reply cedar-policy-validator/src/extensions.rs:48]: Huh. This seems like a mistake. I assume `decimal("1.0")` doesn't parse if the decimal extension isn't enabled?

**Review by khieta — APPROVED** [review]

File: cedar-policy-validator/src/extensions.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (john-h-kastner-aws) [reply cedar-policy-validator/src/extensions.rs:48]: yeah, those don't compile without the feature enabled.

[the remark this answers is not in the document]
Comment (khieta) [reply cedar-policy-validator/src/extensions.rs:48]: Can you make an issue for this?

**Pull request merged**